pkg/gadgets/trace/oomkill/tracer/tracer_test.go

```diff
@@ -0,0 +1,161 @@
+	tracer.Stop()
+	tracer.Stop()
+}
+
+func TestOomkillTracer(t *testing.T) {
+	t.Parallel()
+
+	utilstest.RequireRoot(t)
+
+	_, err := os.Stat("/sys/fs/cgroup/cgroup.controllers")
+	if err != nil {
+		t.Skip("This test requires cgroup v2 to be mounted: %w", err)
+	}
+
+	cgroupPath := "/sys/fs/cgroup/test-oomkill-tracer"
+	memoryLimit := 512 * 1024 // 512 KB.
+
+	// The whole idea of this test is to spawn a child which will allocate so much
+	// memory that it get killed by the OOM killer, hence generating an event.
+	// We cannot do this in the root cgroup, as it could lead to misbehavior from
+	// other test as they would not be able to allocate memory.
+	// So, the whole idea is to create a memory cgroup, set the memory.max limit,
+	// spawn a process, subscribe this child to the memory cgroup and then
+	// allocate the memory until it get killed.
+	err = os.Mkdir(cgroupPath, 0o755)
+	require.NoError(t, err, "creating cgroup path %s", cgroupPath)
+	defer syscall.Rmdir(cgroupPath)
+
+	memoryMaxPath := fmt.Sprintf("%s/memory.max", cgroupPath)
+	file, err := os.OpenFile(memoryMaxPath, os.O_WRONLY, 0)
+	require.NoError(t, err, "opening %s", memoryMaxPath)
+
+	_, err = file.Write([]byte(fmt.Sprintf("%d\n", memoryLimit)))
+	file.Close()
+	require.NoError(t, err, "writing %d to %s", memoryLimit, memoryMaxPath)
+
+	events := []types.Event{}
+	eventCallback := func(event *types.Event) {
+		// normalize
+		event.Timestamp = 0
+
+		events = append(events, *event)
+	}
+
+	sock, err := unix.Socketpair(unix.AF_UNIX, unix.SOCK_DGRAM, 0)
+	require.NoError(t, err, "creating socketpair")
+
+	createTracer(t, &tracer.Config{}, eventCallback)
+
+	childPid, _, errno := syscall.Syscall(syscall.SYS_FORK, 0, 0, 0)
+	require.Equal(t, errno, syscall.Errno(0x0), "spawning child process")
+
+	if childPid == 0 {
+		unix.Recvmsg(sock[1], []byte{}, []byte{}, 0)
+
+		// The child is now part of the memory cgroup, let's allocate some memory.
+		// The memory will be automatically touched by golang as it will initialize
+		// to 0, so no lazy allocation there.
+		// As a result, the child should be OOM killed.
+		// NOLINTNEXTLINE
+		buf := make([]byte, memoryLimit*100)
+
+		t.Fatal("this code should never be reached")
+
+		// Only to make compiler happy.
+		buf[0] = 0
+	}
+
+	procsPath := fmt.Sprintf("%s/cgroup.procs", cgroupPath)
+	procsFile, err := os.OpenFile(procsPath, os.O_WRONLY, 0)
+	require.NoError(t, err, "opening %s", procsPath)
+
+	_, err = procsFile.Write([]byte(fmt.Sprintf("%d\n", childPid)))
+	procsFile.Close()
+	require.NoError(t, err, "writing %d to %s", childPid, procsPath)
+
+	proc, err := os.FindProcess(int(childPid))
+	require.NoError(t, err, "no process with PID %d", childPid)
+
+	// Unblock the child.
+	unix.Sendmsg(sock[0], []byte{}, []byte{}, nil, 0)
+
+	_, err = proc.Wait()
+	require.NoError(t, err, "waiting child with PID %d", childPid)
+
+	// Give some time for the tracer to capture the events
+	time.Sleep(5 * time.Second)
```

Comment:
We should not wait the full 5 seconds if the event arrives earlier.

The `eventCallback()` above can send a message in a channel and here we can `select{}` for either the timeout or the event.